### ISSUE TITLE
Documentation overhaul roadmap

### ISSUE THREAD

**Issue by darsnack** [FluxML/Flux.jl#2105]
**This issue is a WIP. I am starting from the template discussed on call, we need to break this into smaller steps.**

# Flux documentation restructure

It should surprise no one that Flux's current documentation is not very good. It might surprise people that the "whole suite" of docs does have reasonably good coverage—that speaks to the poor organization of the documentation. Despite being written down, most information is an ordeal to find.

With this in mind, we consolidated a list of steps to improve this situation. These steps have been discussed by maintainers and contributors on various Julia platforms, but this issue will now serve as a central tracking for the re-haul.

- [ ] **Restructure the Flux.jl documentation**
      The current doc model was designed at a time when Flux's ethos was to demonstrate how _simple_ neural networks in Julia can be. The current package offers a lot more than a simple list of layers and tracker-based AD. The proposed organization here is a commonly used format (notably in Python fast.ai and our own SciML). See [below](#new-documentation-structure) for a description of this format.
- [ ] **Deprecate website tutorials and model zoo in favor of documentation**
- [ ] **Detailed documentation for NNlib.jl, FluxTraining.jl, etc.**

### New documentation structure

**Home / Getting Started:** Briefly describe the package and quick installation instructions. Provide links to (a) the entry point for an absolute beginner, (b) the ecosystem page, (c) contributing and bug reporting resources.

[**Ecosystem:**](https://fluxml.ai/Flux.jl/stable/ecosystem/) This page is currently way down at the bottom, and it needs to be prominently featured. This should immediately make clear what it is the Flux provides (layers + simple utilities) and what it does not / what it borrows (datasets, data loading, optimizers, schedules, etc.). A diagram like [MLJ's](https://github.com/alan-turing-institute/MLJ.jl/blob/dev/material/MLJ_stack.svg) would be useful.

**Tutorials:** These should be complete, runnable examples that users can use to get started quickly and learn by doing. *The model zoo examples should be tutorials.* [Literate.jl](https://fredrikekre.github.io/Literate.jl/v2/documenter/) offers the ability to generate markdown for consumption by Documenter.jl. Tutorials should leverage this feature.
- **Beginner:** These should be basic tutorials like "taking your first gradient" or "running MNIST." They should rely heavily on the ecosystem packages that a typical beginner would use. For example, use MLDatasets.jl to fetch MNIST instead of rolling your own dataset downloader + type. Writing a data pipeline from scratch provides no value to someone who may not even know what a CNN does. Some existing parts of the Flux docs that fit this area are:
    - ["Overview"](https://fluxml.ai/Flux.jl/stable/models/overview/) which should be renamed to "First steps" or "What is machine learning?"
    - ["Basics"](https://fluxml.ai/Flux.jl/stable/models/basics/) which should be renamed to "Getting started with neural networks" or "What is a layer?"
    - ["Training"](https://fluxml.ai/Flux.jl/stable/training/training/) which probably should be re-written to be more like a tutorial on understanding the basics of a training loop (i.e. remove the docstrings, etc.)
- **Intermediate:** These should be where most model zoo examples should end up. Existing pages that belong here are:
    - ["Regularisation"](https://fluxml.ai/Flux.jl/stable/models/regularisation/)
- **Advanced:** These should be where complex examples with unconventional training schemes end up.

**How-tos:** Unlike tutorials, which are focused on end-to-end examples, how-tos are focused on small bite-sized examples. They should demostrate doing a single thing only. Existing pages here could be:
- ["GPU Support"](https://fluxml.ai/Flux.jl/stable/gpu/)
- ["Saving & Loading"](https://fluxml.ai/Flux.jl/stable/saving/)
- ["Custom model example"](https://fluxml.ai/Flux.jl/stable/models/advanced/#Custom-Model-Example) which could be "Writing custom layers"

**Reference documentation:** These pages explain how certain concepts are designed in Flux from first principles. They are pedagogical like tutorials, but they are not focused on end-to-end examples. Instead of bringing together multiple concepts, reference docs bring explain a single concept in the context of FluxML.
- ["Recurrence"](https://fluxml.ai/Flux.jl/stable/models/recurrence/)
- ["Performance tips"](https://fluxml.ai/Flux.jl/stable/performance/)

**API documentation:** Flux likes to weave docstrings in and out of tutorials which is horrible for finding API references. We should have a single category that is nothing but docstrings organized by package and type.

**Developer documentation:** Flux currently lacks explanations for how things work under the hood. This is a place to explain how all the packages (Functors.jl, NNlib.jl, Zygote.jl, etc.) come together in a gradient call.

**Comment by Saransh-cpp** [comment]
Maybe we can start by segregating the current getting started and tutorials sections into separate folders? We can also add a new "how-to" section with the `Dataloader` example to start things up!

**Comment by mcabbott** [comment]
> ["Overview"](https://fluxml.ai/Flux.jl/stable/models/overview/) which should be renamed to "First steps" or "What is machine learning?"
> ["Basics"](https://fluxml.ai/Flux.jl/stable/models/basics/) which should be renamed to "Getting started with neural networks" or "What is a layer?"

These have already been re-titled, to "Fitting a Straight Line" and "Gradients and Layers". 

> ["Training"](https://fluxml.ai/Flux.jl/stable/training/training/) which probably should be re-written to be more like a tutorial on understanding the basics of a training loop (i.e. remove the docstrings, etc.)

Yes. Doing this is a big job which I didn't want to try in #2069. It may as well be done at the same time as changing to explicit parameters in that section.

> API documentation: Flux likes to weave docstrings in and out of tutorials which is horrible for finding API references. We should have a single category that is nothing but docstrings organized by package and type.

Yes. The book emoji from #2069 are an attempt to do this, without moving too much text in individual files. They aren't great, and a folder separation would probably be better.

One vision for folder structure is:

https://github.com/FluxML/Flux.jl/pull/2069#issuecomment-1273941199

This rolls many existing pages into "Guide".  

The proposal above is to group those into many more different categories, according to whether they are for beginners / experts / developers / how-to / reference. I'm not so sure this will help make things easier to find, IMO too many top-level titles just makes it a guessing game which title the topic of interest is likely to be filed under. 

The separation which does seem clear to me is that "Guide" is documentation about some part of Flux, and "Tutorials" are worked examples of using Flux. Almost every item in https://fluxml.ai/tutorials/ except 1 goes data, architecture, fit... which is great, and clearly a different animal. Learn-by-doing, not by being told. It's fine if several cover the same ground. Whereas "Guide" should try to say each thing once.

**Comment by dorn-gerhard** [comment]
I could contribute on the tutorial side. I am fan of interactive Pluto notebooks, so you could directly play around with some neural networks. I think for a cool tutorial example you should be able to explore the data, see the learning process (training and validation) and check the result (testing)

Are embedded Pluto Notebooks an option for the documentation? 
(like on the [computational thinking course](https://computationalthinking.mit.edu/Fall22/images_abstractions/newton_method/))

**Comment by mcabbott** [comment]
Pluto is great. Do you have in mind something small enough that if it re-runs training repeatedly, waiting for it isn't too much of a pain?

Not quite sure how it could be set up. Hosting live notebooks may be hard, but perhaps something pre-computed like this could work:

https://github.com/JuliaPluto/PlutoSliderServer.jl#3-wip-precomputed-slider-server

Finally, at the moment the docs use Documenter which wants individual pages in as plain markdown files. Do you know if anyone has made Pluto-generated HTML fit in nicely with that?

**Comment by ToucheSir** [comment]
If https://github.com/fonsp/Pluto.jl/pull/985 can be leveraged to be relatively ergonomic for new users (i.e. it's clear why certain cells aren't running and how to enable them), then Pluto could be in the running. That and making sure the generated file is legible enough. Last I checked it does a decent job of not cluttering source code, but didn't check how markdown blocks look.

**Comment by dorn-gerhard** [comment]
regarding disabled cells, it's now way easier to find the reason why a cell is disabled, you can simple jump to that cell.

What I have in mind for a sample notebook is one, where you can draw digits and and the notebook tries to recognize it.
Using some user input, like [Connor did](https://cotangent.dev/how-to-make-custom-pluto-ui-components/).

Not 100% sure how to integrate Pluto notebooks into Documenter but I will ask ;)

**Comment by CarloLucibello** [comment]
Closing — the restructure this roadmap called for has happened. `docs/src` is now organised as `guide/` (quickstart, models, training, GPU, performance, saving, Reactant), `reference/` (layers, losses, optimisers, gradients, data, utilities), `tutorials/` (linear regression, logistic regression, custom layers, model zoo) and `ecosystem.md`.

Remaining documentation gaps are much better tracked as individual issues than as one long-lived checklist.
